### DISCUSSION TITLE
Mesh extrusion and simplex conversion

### DISCUSSION THREAD

**Discussion by CodeFloh** [tataratat/gustaf#109] [Ideas]
I'm gonna create one issue even though these aspects could be treated separately. I'm interested in the following functionality:

- Extrude a two-dimensional FE mesh to a three-dimensional mesh with a specified number of layers and thickness.
- Convert non-simplex to simplex meshes.

These two operations do not necessarily need to be separate. The following specific conversions are useful:

- [x] 2D quadrilateral mesh to 2D triangle mesh
- [ ] 3D hexahedron mesh to 3D tetrahedron mesh
- [ ] 2D triangle mesh to 3D tetrahedron mesh

All these operations need to conserve the boundary information. Extrusion operations should add an additional boundary.

Answers to the following questions could help:

- How would these operations be carried out in gustav?
- Where in gustav can the functions be found?
- Where in gustaf should this functionality be implemented?

**Comment by j042** [comment]
_2D quadrilateral mesh to 2D triangle mesh_ -> `Mesh.tri_mesh`
_3D hexahedron mesh to 3D tetrahedron mesh_ -> As discussed, this would be not exact transformation (should give warning!), but I like the idea of providing simple 1 hexa to 6 tetra. and once we have a fancier "tetrahedralizer", we can extend this. 
_2D triangle mesh to 3D tetrahedron mesh_ -> `gustav.mesh_shapes.extrude_volume`



> * Where in gustaf should this functionality be implemented?

=>`gustaf/create/faces.py` & `gustaf/create/volumes.py` ?

**Comment by j042** [comment]
@LinusKnecht @CodeFloh check out [tetgenpy](https://github.com/tataratat/tetgenpy). It should provide stable and fast solutions for generating tetrahedrons and covers all the use-case you've stated above. We can discuss implementation details if you are interested

**Reply by LinusKnecht** [reply]
It looks quite nice to me at a first view. The conversion to gustaf and vice versa seems to work as well.
